okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,142 @@
+import okhttp3.Dns;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.internal.platform.Platform;
+import okio.ByteString;
+
+/**
+ * DNS over HTTPS implementation.
+ *
+ * Implementation of https://tools.ietf.org/html/draft-ietf-doh-dns-over-https-09
+ *
+ * <blockquote>A DNS API client encodes a single DNS query into an HTTP request
+ * using either the HTTP GET or POST method and the other requirements
+ * of this section.  The DNS API server defines the URI used by the
+ * request through the use of a URI Template.</blockquote>
+ */
+public class DnsOverHttps implements Dns {
+  public static final MediaType DNS_MESSAGE = MediaType.parse("application/dns-message");
+  public static final MediaType UDPWIREFORMAT = MediaType.parse("application/dns-udpwireformat");
+  private final OkHttpClient client;
+  private final HttpUrl url;
+  private final boolean includeIPv6;
+  private final boolean post;
+  private final MediaType contentType;
+
+  public DnsOverHttps(OkHttpClient client, HttpUrl url, @Nullable Dns bootstrapDns,
+      boolean includeIPv6, String method, MediaType contentType) {
+    this.client = bootstrapDns != null ? client.newBuilder().dns(bootstrapDns).build() : client;
+    this.url = url;
+    this.includeIPv6 = includeIPv6;
+    if (!method.equals("GET") && !method.equals("POST")) {
+      throw new UnsupportedOperationException("Only GET and POST Supported");
+    }
+    this.post = method.equals("POST");
+    this.contentType = contentType;
+  }
+
+  public HttpUrl getUrl() {
+    return url;
+  }
+
+  public boolean isPost() {
+    return post;
+  }
+
+  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
+    try {
+      ByteString query = DnsRecordCodec.encodeQuery(hostname, includeIPv6);
+
+      Request request = buildRequest(query);
+      Response response = executeRequest(request);
+
+      return readResponse(hostname, response);
+    } catch (UnknownHostException uhe) {
+      throw uhe;
+    } catch (Exception e) {
+      UnknownHostException unknownHostException = new UnknownHostException(hostname);
+      unknownHostException.initCause(e);
+      throw unknownHostException;
+    }
+  }
+
+  private Response executeRequest(Request request) throws IOException {
+    // cached request
+    if (!post && client.cache() != null) {
+      Request cacheRequest = request.newBuilder().cacheControl(CacheControl.FORCE_CACHE).build();
+
+      Response response = client.newCall(cacheRequest).execute();
+
+      if (response.isSuccessful()) {
+        return response;
+      }
+    }
+
+    return client.newCall(request).execute();
+  }
+
+  private List<InetAddress> readResponse(String hostname, Response response) throws Exception {
+    if (response.cacheResponse() == null && response.protocol() != Protocol.HTTP_2) {
```

Comment:
oh interesting. We don’t remember the protocol for cached responses!

Reply:
Do cached responses even have a protocol?

Reply:
we could store it? Probably not worth the effort.